### PR TITLE
[plog-client] add a PlogClient

### PR DESCRIPTION
**Pull request by hardproblems**
This is the PlogClient that Xiang wrote and is currently used by our java services. I didn't want to change this because it is already known to work in production.

cc @alexism

### CODE REVIEW

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      chunks.add(Arrays.copyOfRange(messageBytes, startIndex, messageBytes.length));
+    }
+    return chunks;
+  }
+
+  private void openSocket() {
+    try {
+      socket = new DatagramSocket();
+    } catch (SocketException e) {
+      log.error("Cannot reopen socket", e);
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: `reopen` -> `open`

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ * plogClient.send("My hovercraft is full of eels.");
+ * ```
+ * You can configure the client at initialization by passing these options:
+ *
+ * + host - The host of the Plog process (e.g., 'localhost')
+ * + port - The port on which Plog is listening (e.g., 23456)
+ * + chunkSize - The maximum payload size for multipart datagrams (e.g., 64,000)
+ */
+@Slf4j
+public class PlogClient {
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: should `PlogClient` be a `Closeable`? and provide a `close` method that closes the socket

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  private final InetAddress address;
+
+  private final int port;
+
+  private final int chunkSize;
+
+  private DatagramSocket socket;
+
+  public PlogClient(String host, int port, int chunkSize) {
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: some default value for `chunkSize` would be helpful

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    } catch (SocketException e) {
+      log.error("Cannot reopen socket", e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private void sendToSocket(byte[] chunk) {
+    DatagramPacket packet = new DatagramPacket(chunk, chunk.length, address, port);
+    try {
+      log.debug("Sending {} to UDP port {}", chunk, port);
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: this is potentially quite verbose. use `trace` instead?

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      socket.send(packet);
+    } catch (IOException e) {
+      socket.close();
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: log an error message?

File: plog-client/src/test/groovy/com/airbnb/plog/client/MultipartMessageTest.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.airbnb.plog.client;
+
+import java.nio.charset.Charset;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertArrayEquals;
+
+public class MultipartMessageTest {
```
Comment (alexism) [inline plog-client/src/test/groovy/com/airbnb/plog/client/MultipartMessageTest.java]: some unit tests on `chunkMessage()` would be good too

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 2 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: Yea that makes sense. Will add.

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 1 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: Good catch!

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 4 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: Will change

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 3 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: Where should I specify a default chunkSize?

Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: nevermind will add a constant and another constructor with no chunkSize required

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  private DatagramSocket socket;
+
+  public PlogClient(String host, int port) {
+    this(host, port, DEFAULT_CHUNK_SIZE);
+  }
+
+  public PlogClient(String host, int port, int chunkSize) {
+    Preconditions.checkNotNull(host, "host cannot be null!");
+    Preconditions.checkArgument(port > 1024 && port < 65536, "Must provide a valid port number!");
+    Preconditions.checkArgument(chunkSize < 65507, "Maximum UDP data length is 65507 bytes!");
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: will a chunk size of `65507` be ok? This value does not take the Plog header size into account

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    } catch (IOException e) {
+      log.error("Error sending packet!", e);
+      socket.close();
+      openSocket();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    socket.close();
```
Comment (alexism) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: `socket` might be `null` if an exception was thrown in `sendToSocket` `catch` clause.

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 7 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: good catch, will reduce by 24 bytes to 65483

File: plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java (hunk 8 above)
Comment (hardproblems) [inline plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java]: good point! will check.

File: plog-client/src/test/groovy/com/airbnb/plog/client/MultipartMessageTest.java (hunk 6 above)
Comment (hardproblems) [inline plog-client/src/test/groovy/com/airbnb/plog/client/MultipartMessageTest.java]: added test in PlogClientTest

**Pull request merged**